fall back to eager on `NotImplementedError`

ev-br opened the pull request:
Follow-up to https://github.com/pytorch/pytorch/pull/107710:

Help  dynamo fall back to eager when compiling unimplemented numpy constructs:

- arrays of strings
- (arg){min, max} for complex types
- various arguments typed as NotImplemented (`np.ones(4, order="F")` etc)
- numpy functions which torch._numpy does not implement

To test, run (we do not implement arrays of strings)

```
import torch
import numpy as np

@torch.compile(fullgraph=False)
def fn():
    return np.asarray(["L", "U"])
```

and observe it compiles with fullgraph=False and fails with fullgraph=True

Fixes https://github.com/pytorch/pytorch/issues/107970

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @gujinghui @PenghuiCheng @jianyuh @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @anijain2305

ev-br committed:
MAINT: return NotImplemented in reductions on complex arrays

NumPy allows that but PyTorch does not, so fall back to eager

ev-br committed:
MAINT: fall back to eager for all args types as NotImplementedType

ev-br committed:
MAINT: fall back to eager on NumPy names which torch._numpy does not implement

lezcano reviewed (commented):
Please  audit all the `NotImplementedError`s in the code. At the moment, we have a few others that fall outside of the `normalizer`, for example in `_ndarray`.

lezcano commented on torch/_numpy/__init__.py:
```diff
@@ -26,3 +26,10 @@
 
 False_ = False
 True_ = True
+
+
+def __getattr__(name):
+    # is only called for names not found via a normal lookup (cf PEP 562)
+    # returning NotImplemented allows dynamo to fall back to eager,
+    # meaning the name will be searched on the NumPy namespace.
+    return NotImplemented
```
Also, cool trick, but I don't think it's necessary. dynamo already knows how to fallback in this case.

ev-br committed:
TST: fix tests

ev-br commented:
This PR should be failing CI left and right because of a syntax error (which I fixed locally bit did not push the fix). So can be used as a canary on a coalmine to check that torch_np tests actually run.

ev-br committed:
a mess to make np.asarray(["L", "R"]) fall back to eager

ev-br committed:
TST: add a test

ev-br committed:
MAINT: simplify the logic, raise/catch NotImplementedError directly

- in tnp.asarray, a failure to convert to tensor triggers NotImplementedError
  regardless of the reason
- in dynamo, run_node, convert NotImplementedError into
  UnsupportedFakeTensorException, which allows the fallback to eager

ev-br committed:
MAINT: raise NotImplementedError from _numpy reductions

ev-br committed:
lint

ev-br committed:
fix assert_equal for strings

ev-br commented on torch/_numpy/_normalizations.py:
```diff
@@ -193,23 +193,29 @@ def wrapped(*args, **kwds):
+                    name: maybe_normalize(arg, params[name]) if name in params else arg
+                    for name, arg in kwds.items()
+                }
+            except (NotImplementedError, RuntimeError, ValueError):
```
This try-except is gone in the revised version.

ev-br commented:
Pushed an update to simplify the logic:

- in tnp.asarray, a failure to convert to tensor triggers NotImplementedError regardless of the reason
- `run_node`, convert catch NotImplementedError and convert it into  UnsupportedFakeTensorException

WDYT @lezcano?

ev-br commented on torch/_dynamo/utils.py:
```diff
@@ -1406,6 +1409,10 @@ def run_node(tracer, node, args, kwargs, nnmodule):
         elif op == "placeholder":
             assert "example_value" in node.meta
             return node.meta["example_value"]
+    except Unsupported:
```
The updated version catches `NotImplementedError`, so the logic is basically: graph break on NotImplementedError, error out on all other exceptions.
And this seems to be the place which dispatches various call_function, call_method and other operations.

lezcano commented on torch/_numpy/_reductions_impl.py:
```diff
@@ -71,6 +71,9 @@ def argmax(
     keepdims: KeepDims = False,
 ):
+    if a.is_complex():
+        raise NotImplementedError
```
Raise an instance. Same below.

lezcano reviewed (commented):
This looks much better. That being said, I'm tagging Ed see if it's alright with him, or there's a better approach. Let's wait for his review.

ev-br committed:
MAINT: raise instances not classes of *Error

lezcano commented:
@ev-br the solution seems to break other tests. I think there must be another way of doing things, this one looks too hacky. Tagging @voznesenskym and @anijain2305 for ideas

ev-br committed:
reuse unimplemented(...), make test more explicit

ev-br committed:
lint

ev-br committed:
TST: adjust tests for XPASS

lezcano commented on test/torch_np/numpy_tests/lib/test_function_base.py:
```diff
@@ -3434,7 +3434,6 @@ def test_basic(self):
         a = np.array([0.0444502, 0.141249, 0.0463301])
         assert_equal(a[-1], np.median(a))
 
-    @pytest.mark.xfail(reason="median: scalar output vs 0-dim")
```
Why did all these tests use to fail and now they pass?

ev-br committed:
Revert "TST: adjust tests for XPASS"

This reverts commit d452632b6a07aaf526bf923ef9e95eccc3b77092.

ev-br committed:
BUG: fix assert_equal

ezyang reviewed (approved):
Seems fine. You weren't relying on NotImplemented sentinel to do multi dispatch for binary ops right?

ezyang commented:
guess you gotta hose down some newly uncovered bugs now

lezcano commented:
Nah, we added the `NotImplemented` as a workaround, but exceptions are better.

ev-br committed:
TST/XXX: xfail a sparse test

cf https://github.com/pytorch/pytorch/issues/108667

ev-br commented:
The last commit xfails the sparse test from https://github.com/pytorch/pytorch/issues/108667

lezcano commented on test/test_sparse.py:
```diff
@@ -591,6 +591,7 @@ def test_sparse_bool(self, device, dtype):
         b = a.to_sparse().to_dense()
         self.assertEqual(a, b)
 
+    @unittest.expectedFailure         # cf https://github.com/pytorch/pytorch/issues/108667
```
Nope. We have a decorator that xfails if run with dynamo. That's what you want.

amjames replied:
FWIW I did try this out locally and if that error were to change from `NotImplementedError` to something else, this all works fine.

ev-br replied:
https://github.com/pytorch/pytorch/pull/107863/commits/8eabd14675241dd0a27b777ae8dd270b424c13ba `skipIfTorchDynamo`  as used elsewhere in this test suite

lezcano replied:
@amjames yes, this is a small bug in TorchDynamo, that it should create a break properly and fallback to eager but it doesn't. If anyone from sparse wants to have a knock at this one it should be a good introductory PR.

ev-br committed:
TST: skipIfTorchDynamo

ev-br committed:
TST: skip a pinned memory test under dynamo

lezcano reviewed (approved):
We should investigate at some point why the tests were passing before and not breaking now, but that can be done at a later time.

lezcano commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot rebase

lezcano commented:
@pytorchbot merge

ev-br commented:
@pytorchbot merge -i

ev-br commented:
@pytorchbot merge -i

atalman commented:
will be included in 2.2
